lib/scripts/migrate-redis.sh

```diff
@@ -0,0 +1,166 @@
+NEW_REDIS_PASSWORD=$(cloud-platform decode-secret -n $NEW_NAMESPACE -s laa-check-client-qualifies-elasticache-instance-output --skip-version-check | jq '.data.auth_token' | tr -d '"')
+
+if [[ ! $OLD_REDIS_HOST ]]; then
+  echo "Could not retrieve old namespace hostname"
+  exit 1
+fi
+
+if [[ ! $OLD_REDIS_PASSWORD ]]; then
+  echo "Could not retrieve old namespace password"
+  exit 1
+fi
+
+if [[ ! $NEW_REDIS_HOST ]]; then
+  echo "Could not retrieve new namespace hostname"
+  exit 1
+fi
+
+if [[ ! $NEW_REDIS_PASSWORD ]]; then
+  echo "Could not retrieve new namespace password"
+  exit 1
+fi
+
+
+STANDARD_REDIS_PORT=6379
+OLD_REDIS_LOCAL_PORT=6380
+NEW_REDIS_LOCAL_PORT=6381
+PORT_FORWARD_POD_NAME=redis-port-forward-pod
+
+# Create port forwarding pods
+create_pods() {
+  echo "3/11 Creating port forwarding pod in old namespace..."
+  kubectl \
+    -n $OLD_NAMESPACE \
+    run $PORT_FORWARD_POD_NAME \
+    --image=ministryofjustice/port-forward \
+    --port=$STANDARD_REDIS_PORT \
+    --env="REMOTE_HOST=$OLD_REDIS_HOST" \
+    --env="LOCAL_PORT=$STANDARD_REDIS_PORT" \
+    --env="REMOTE_PORT=$STANDARD_REDIS_PORT" || { echo "Failed to create pod in old namespace"; return 1; }
+
+  echo "4/11 Creating port forwarding pod in new namespace..."
+  kubectl \
+    -n $NEW_NAMESPACE \
+    run $PORT_FORWARD_POD_NAME \
+    --image=ministryofjustice/port-forward \
+    --port=$STANDARD_REDIS_PORT \
+    --env="REMOTE_HOST=$NEW_REDIS_HOST" \
+    --env="LOCAL_PORT=$STANDARD_REDIS_PORT" \
+    --env="REMOTE_PORT=$STANDARD_REDIS_PORT" || { echo "Failed to create pod in new namespace"; return 1; }
+
+  echo "5/11 Waiting 10 seconds for pods to be ready..."
+  sleep 10
+  return 0
+}
+
+# Start running the port forwarding ports
+run_pods() {
+  echo "6/11 Starting port forwarding pod in old namespace..."
+  kubectl -n $OLD_NAMESPACE port-forward $PORT_FORWARD_POD_NAME $OLD_REDIS_LOCAL_PORT:$STANDARD_REDIS_PORT &
+
+  OLD_PORT_FORWARD_PID=$!
+
+  echo "7/11 Waiting 3 seconds to see if port forwarding fails early in old namespace..."
+  sleep 3
+  if [[ ! $(ps -p $OLD_PORT_FORWARD_PID -o pid=) ]]; then
+    echo "Port forwarding in old namespace failed"
+    return 1
+  fi
+
+  echo "8/11 Starting port forwarding pod in new namespace..."
+  kubectl -n $NEW_NAMESPACE port-forward $PORT_FORWARD_POD_NAME $NEW_REDIS_LOCAL_PORT:$STANDARD_REDIS_PORT &
+
+  NEW_PORT_FORWARD_PID=$!
+
+  echo "9/11 Waiting 3 seconds to see if port forwarding fails early in new namespace..."
+  sleep 3
+  if [[ ! $(ps -p $NEW_PORT_FORWARD_PID -o pid=) ]]; then
+  echo "Port forwarding in new namespace failed"
+    return 1
+  fi
+
+  return 0
+}
+
+# Run the RIOT tool that will push data from old to new in batches
+run_riot() {
+  echo "10/11 Running riot..."
+  riot -h localhost -p $OLD_REDIS_LOCAL_PORT --pass $OLD_REDIS_PASSWORD --tls --tls-verify NONE \
+      replicate \
+      -h localhost -p $NEW_REDIS_LOCAL_PORT --pass $NEW_REDIS_PASSWORD --tls --tls-verify NONE --batch 10000 \
```

Comment:
looks good, I'm not too familiar with Riot - just so I'm understanding, if the migration fails here midway through, the redis pods get deleted anyway. So we wouldn't end up in a situation with half of the redis data imported. Is that right?

Reply:
Ah, not exactly. What get deleted are the 'port forwarding' pods - the things that allow us to access redis from our local machines.

So in the worst case scenario, riot runs and gets half of the redis data migrated, and then fails out. Everything gets tidied up, the port forwarding pods get deleted, and when the dust clears the new namespace has half of the data from the old namespace. I can't see a riot option that does the whole transaction atomically, so if this happens, as it could, our recourse would have to be to use the `redis-cli` to do a FLUSHALL on the new namespace and try again or similar. Which isn't ideal, but I can't see a better way, and I think it's manageable if it comes to it.